### PR TITLE
Ok did power ratings!

### PR DESCRIPTION
**Pull request by FavouriteDragon**
I still need to make the normal abilities affected by power ratings though
But that's low priority, levels and xp come first.

### CODE REVIEW

File: src/main/java/com/crowsofwar/avatar/common/bending/air/SlipstreamPowerModifier.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import com.crowsofwar.avatar.common.data.PowerRatingModifier;
+import com.crowsofwar.avatar.common.data.ctx.BendingContext;
+
+public class SlipstreamPowerModifier extends PowerRatingModifier {
+    @Override
+    public double get(BendingContext ctx) {
+        AbilitySlipstream slipstream = new AbilitySlipstream();
+        BendingData data = ctx.getData();
+        AbilityData abilityData = data.getAbilityData(slipstream);
+        setTicks(20+(20*abilityData.getLevel()));
```
Comment (CrowsOfWar) [inline src/main/java/com/crowsofwar/avatar/common/bending/air/SlipstreamPowerModifier.java]: The setTicks method sets the amount of time left for the power modifier. If you're resetting the countdown every time the modifier amount is calculated, you are essentially extending the time every time the the power rating is calculated. This is problematic since you wouldn't want to keep extending the time the modifier is active.

I think you're still trying to customize how long the modifier is applied right?  Just call setTicks when you first apply a new modifier (I've commented a good place to do this)

Thanks!

File: src/main/java/com/crowsofwar/avatar/common/bending/earth/RestorePowerModifier.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import com.crowsofwar.avatar.common.data.PowerRatingModifier;
+import com.crowsofwar.avatar.common.data.ctx.BendingContext;
+
+public class RestorePowerModifier extends PowerRatingModifier {
+    @Override
+    public double get(BendingContext ctx) {
+        AbilityRestore restore = new AbilityRestore();
+        BendingData data = ctx.getData();
+        AbilityData abilityData = data.getAbilityData(restore);
+        setTicks(20+(20*abilityData.getLevel()));
```
Comment (CrowsOfWar) [inline src/main/java/com/crowsofwar/avatar/common/bending/earth/RestorePowerModifier.java]: See the comments for SlipstreamPowerModifier- setTicks shouldn't be used here

File: src/main/java/com/crowsofwar/avatar/common/bending/fire/AbilityPurify.java [hunk 3]
```diff
@@ -40,6 +39,7 @@ public void execute(AbilityContext ctx) {
                 entity.addPotionEffect(new PotionEffect(MobEffects.STRENGTH, 100));
                 entity.addPotionEffect(new PotionEffect(MobEffects.SPEED, 100));
                 data.getAbilityData("purify").addXp(xp-0.5F);
+                chi *= 1.5f;
```
Comment (CrowsOfWar) [inline src/main/java/com/crowsofwar/avatar/common/bending/fire/AbilityPurify.java]: Theres a subtle but annoying problem here- the chi variable is being modified after its actually used. So the new value isn't used at all. This is kinda confusing so let's go through this step by step and hopefully my bad explanations will be ok enough...

1. The chi is initialized to the config value
2. Consume chi based on the chi variable. At this point, the chi variable is unaffected by XP! So the amount consumed is only the config value, not having to do with xp
3. Now the chi variable is modified based on XP, but chi isn't used any more so it doesn't really matter.

So maybe the xp related calculations on chi variable should happen BEFORE the chi is actually consumed.

Comment (CrowsOfWar) [reply src/main/java/com/crowsofwar/avatar/common/bending/fire/AbilityPurify.java]: Please let me know if this was hard to follow! I am bad at explaining things.

File: src/main/java/com/crowsofwar/avatar/common/bending/fire/AbilityPurify.java [hunk 4]
```diff
@@ -57,16 +58,19 @@ public void execute(AbilityContext ctx) {
...
             if (data.getAbilityData("purify").isMasterPath(AbilityData.AbilityTreePath.SECOND)) {
                 entity.addPotionEffect(new PotionEffect(MobEffects.NIGHT_VISION, 200));
                 entity.addPotionEffect(new PotionEffect(MobEffects.STRENGTH, 200, 1));
                 entity.addPotionEffect(new PotionEffect(MobEffects.SPEED, 200, 1));
                 entity.addPotionEffect(new PotionEffect(MobEffects.GLOWING, 200));
+                chi *= 2.5f;
             }
 
         }
+        data.getPowerRatingManager(getBendingId()).addModifier(new PurifyPowerModifier());
```
Comment (CrowsOfWar) [inline src/main/java/com/crowsofwar/avatar/common/bending/fire/AbilityPurify.java]: Here is where to call setTicks. This is where the power modifer is created, which is when its duration should be set.

File: src/main/java/com/crowsofwar/avatar/common/bending/air/SlipstreamPowerModifier.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+        return 10+(3*abilityData.getLevel());
```
Comment (CrowsOfWar) [inline src/main/java/com/crowsofwar/avatar/common/bending/air/SlipstreamPowerModifier.java]: Cool, you're making the power modifier increase when the player levels up. That's a great little detail! Details go a long way.

File: src/main/java/com/crowsofwar/avatar/common/bending/earth/AbilityRestore.java [hunk 6]
```diff
@@ -28,36 +29,44 @@ public void execute(AbilityContext ctx) {
...
             float xp = SKILLS_CONFIG.blockPlaced;
+
             entity.addPotionEffect(new PotionEffect (MobEffects.INSTANT_HEALTH, 1));
             data.getAbilityData("restore").addXp(xp);
+
             if (abilityData.getLevel()==1) {
                 entity.addPotionEffect(new PotionEffect(MobEffects.INSTANT_HEALTH, 1));
                 entity.addPotionEffect(new PotionEffect(MobEffects.RESISTANCE, 100));
                 data.getAbilityData("restore").addXp(xp);
+                chi *= 1.5f;
```
Comment (CrowsOfWar) [inline src/main/java/com/crowsofwar/avatar/common/bending/earth/AbilityRestore.java]: Theres a subtle but annoying problem here- the chi variable is being modified after its actually used. So the new value isn't used at all. This is kinda confusing so let's go through this step by step and hopefully my bad explanations will be ok enough...

1. The chi is initialized to the config value
2. Consume chi based on the chi variable. At this point, the chi variable is unaffected by XP! So the amount consumed is only the config value, not having to do with xp
3. Now the chi variable is modified based on XP, but chi isn't used any more so it doesn't really matter.

So maybe the xp related calculations on chi variable should happen BEFORE the chi is actually consumed.

Please let me know if this was hard to follow! I am bad at explaining things.

File: src/main/java/com/crowsofwar/avatar/common/bending/water/AbilityCleanse.java [hunk 7]
```diff
@@ -26,37 +24,45 @@ public void execute(AbilityContext ctx) {
...
+
+            entity.addPotionEffect(new PotionEffect(MobEffects.REGENERATION, 100));
             data.getAbilityData("cleanse").addXp(xp);
+
             if (abilityData.getLevel() == 1) {
                 entity.addPotionEffect(new PotionEffect(MobEffects.REGENERATION, 200));
-                entity.addPotionEffect(new PotionEffect(MobEffects.WATER_BREATHING, 200));
+                entity.addPotionEffect(new PotionEffect(MobEffects.WATER_BREATHING, 100));
                 data.getAbilityData("cleanse").addXp(xp);
+                chi *= 1.5f;
```
Comment (CrowsOfWar) [inline src/main/java/com/crowsofwar/avatar/common/bending/water/AbilityCleanse.java]: It's cool how chi cost increases when the ability becomes more powerful! I bet this will be an interesting mechanic. Maybe I'll have to think of some new strategies!

File: src/main/java/com/crowsofwar/avatar/common/bending/water/CleansePowerModifier.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+import com.crowsofwar.avatar.common.data.AbilityData;
+import com.crowsofwar.avatar.common.data.BendingData;
+import com.crowsofwar.avatar.common.data.PowerRatingModifier;
+import com.crowsofwar.avatar.common.data.ctx.BendingContext;
+
+public class CleansePowerModifier extends PowerRatingModifier {
+    @Override
+    public double get(BendingContext ctx) {
+        AbilityCleanse cleanse = new AbilityCleanse();
```
Comment (CrowsOfWar) [inline src/main/java/com/crowsofwar/avatar/common/bending/water/CleansePowerModifier.java]: By the way, just nitpicking again, but technically this wouldn't be needed. Plus, Ability objects generally should only be made once per ability. If this makes an error, it's easy to fix down below vvv

File: src/main/java/com/crowsofwar/avatar/common/bending/water/CleansePowerModifier.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+        BendingData data = ctx.getData();
+        AbilityData abilityData = data.getAbilityData(cleanse);
```
Comment (CrowsOfWar) [inline src/main/java/com/crowsofwar/avatar/common/bending/water/CleansePowerModifier.java]: It's recommended to use the ability name as the parameter, instead of an instance of the ability. It's simpler that way:

data.getAbilityData("cleanse")    <--- using the ability name, instead of an Ability object as the parameter

File: src/main/java/com/crowsofwar/avatar/common/entity/EntityCloudBall.java [hunk 10]
```diff
@@ -57,6 +49,9 @@ public void entityInit() {
     public void onUpdate() {
         super.onUpdate();
         setBehavior((CloudburstBehavior) getBehavior().onUpdate(this));
+        if (ticksExisted >= 250) {
+            this.setDead();
+        }
```
Comment (CrowsOfWar) [inline src/main/java/com/crowsofwar/avatar/common/entity/EntityCloudBall.java]: Thanks for fixing that trello issue!

File: src/main/java/com/crowsofwar/avatar/common/bending/air/SlipstreamPowerModifier.java [hunk 11]
```diff
@@ -0,0 +1,19 @@
+package com.crowsofwar.avatar.common.bending.air;
+
+import com.crowsofwar.avatar.common.data.AbilityData;
+import com.crowsofwar.avatar.common.data.BendingData;
+import com.crowsofwar.avatar.common.data.PowerRatingModifier;
+import com.crowsofwar.avatar.common.data.ctx.BendingContext;
+
+public class SlipstreamPowerModifier extends PowerRatingModifier {
```
Comment (CrowsOfWar) [inline src/main/java/com/crowsofwar/avatar/common/bending/air/SlipstreamPowerModifier.java]: Nice power modifier class!!

File: src/main/java/com/crowsofwar/avatar/common/bending/air/SlipstreamPowerModifier.java [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    @Override
+    public double get(BendingContext ctx) {
+        AbilitySlipstream slipstream = new AbilitySlipstream();
+        BendingData data = ctx.getData();
+        AbilityData abilityData = data.getAbilityData(slipstream);
+        setTicks(20+(20*abilityData.getLevel()));
+        return 10+(3*abilityData.getLevel());
+
+        }
+    }
```
Comment (CrowsOfWar) [inline src/main/java/com/crowsofwar/avatar/common/bending/air/SlipstreamPowerModifier.java]: Just nitpicking here, but shouldn't these curly braces be formatted a bit differently? Thanks lol

File: src/main/java/com/crowsofwar/avatar/common/bending/water/AbilityCleanse.java [hunk 13]
```diff
@@ -26,37 +24,45 @@ public void execute(AbilityContext ctx) {
...
                     }
                           if (data.getAbilityData("cleanse").isMasterPath(AbilityData.AbilityTreePath.SECOND)) {
                             entity.addPotionEffect(new PotionEffect(MobEffects.REGENERATION, 200, 2));
                             entity.addPotionEffect(new PotionEffect(MobEffects.WATER_BREATHING, 200, 1));
                             entity.addPotionEffect(new PotionEffect(MobEffects.SATURATION, 200));
+                              chi *= 2.5f;
+                    }
+
+                }
+                data.getPowerRatingManager(getBendingId()).addModifier(new CleansePowerModifier());
```
Comment (CrowsOfWar) [inline src/main/java/com/crowsofwar/avatar/common/bending/water/AbilityCleanse.java]: Here is also where to call setTicks:

CleansePowerModifier modifier = new CleansePowerModifier();
modifier.setTicks( etc );
data.getPowerRatingManager(getBendingId()).addModifier(modifier);

File: src/main/java/com/crowsofwar/avatar/common/config/ConfigSkills.java [hunk 14]
```diff
@@ -38,6 +38,7 @@ private ConfigSkills() {}
 		blockKill = 4f,
 		airJump = 2f,
 		airGustHit = 4f,
+		buffUsed = 2f,
```
Comment (CrowsOfWar) [inline src/main/java/com/crowsofwar/avatar/common/config/ConfigSkills.java]: good job adding the config value!!

File: src/main/java/com/crowsofwar/avatar/common/bending/fire/AbilityPurify.java (hunk 3 above)
Comment (FavouriteDragon) [reply src/main/java/com/crowsofwar/avatar/common/bending/fire/AbilityPurify.java]: Ok i thought that might be problem but I was too lazy to go and edit the chi before the consume chi part. I guess I'll do that now!

File: src/main/java/com/crowsofwar/avatar/common/bending/water/AbilityCleanse.java (hunk 13 above)
Comment (FavouriteDragon) [reply src/main/java/com/crowsofwar/avatar/common/bending/water/AbilityCleanse.java]: ok thanks!

Comment (FavouriteDragon) [reply src/main/java/com/crowsofwar/avatar/common/bending/water/AbilityCleanse.java]: I had to do this:
data.getPowerRatingManager(getBendingId()).addModifier(new SlipstreamPowerModifier());
instead of using modifier cause it didn't like it

File: src/main/java/com/crowsofwar/avatar/common/bending/water/CleansePowerModifier.java (hunk 9 above)
Comment (FavouriteDragon) [reply src/main/java/com/crowsofwar/avatar/common/bending/water/CleansePowerModifier.java]: ok doing that now!

**Comment by FavouriteDragon** [comment]
Ok please don't merge yet I'm still cleaning up :D

**Comment by FavouriteDragon** [comment]
Ok done!
